Tax range

LewisAJones opened the pull request:
# Function name
tax_range_time() and tax_range_geo()

## Description

tax_range_time() 
A function to calculate the temporal range of fossil taxa.

tax_range_geo()
A function to calculate the geographic range of fossil taxa. The function can calculate geographic range in four ways: convex hull, latitudinal range, maximum great circle distance, and the number of occupied equal-area hexagonal grid cells.

## Motivation and Context

These functions allow the user to calculate common metrics used in palaeobiology (temporal and spatial ranges). Although their use is common, no standard implementation exists. Their inclusion in the package will allow researchers to avoid re-inventing the wheel and provides readable/documented code for how these metrics are calculated.

## Checklist before requesting a review
- [x] I have read palaeoverse's [standards and structure](https://github.com/palaeoverse-community/palaeoverse/raw/main/man/tutorials/structure-and-standards.pdf) 
- [x] I have performed a self-review of my code.
- [x] I have added function documentation.
- [x] I have provided sufficient examples of implementation.
- [x] If it is a core feature, I have added thorough tests.

Closes #28

bethany-j-allen commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+  }
+
+  if (any(is.na(occdf[,name]))) {
+    stop("The `name` column contains NA values")
```
It should maybe be mentioned somewhere in the documentation that these should be filtered out beforehand.

bethany-j-allen commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+    # If the user has entered a non-valid term for the "method" argument,
+    # generate an error and warn the user.
+    stop("Invalid `method`. Choose either:
+  'con', 'lat', 'gcd', and 'occ'.")
```
```suggestion
  'con', 'lat', 'gcd', or 'occ'.")
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+             pch = 20,
+             col = temp_df$taxa_id)
+      axis_geo(side = 1, intervals = "periods")
+      title(xlab = "Time (Ma)", line = 4)
```
For the plot (but also maybe for the table?), it would be nice to be able to sort by time, or at least that's how this information is typically displayed in figures.

LewisAJones replied:
Great suggestion! I have added an additional argument (by), which allows the user to specify how to sort both the output and plot. Options include FAD, LAD and name.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    # Return dataframe
+    return(temp_df)
```
Might be useful to add a column with the number of occurrences?

LewisAJones replied:
Yes, why not! Integrated.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @importFrom graphics points
+#' @examples
+#' # Grab internal data
+#' occdf <- tetrapods
```
As above, I think it would be best to clean this for either species or genera in the example.

LewisAJones replied:
Updated to use genus-level assignments. Requires some NA removal before, but that is fine!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' Calculate the temporal range of fossil taxa
+#'
+#' A function to calculate the temporal range of fossil taxa.
```
```suggestion
#' A function to calculate the temporal range of fossil taxa from occurrence data.
```

LewisAJones replied:
Added!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#'
+#' @param occdf \code{dataframe}. A dataframe of fossil occurrences. The
+#' dataframe should contain at least three columns: names of taxa (`character`)
+#' and the maximum (`numeric`) and minimum (`numeric`) age which constrain the
```
```suggestion
#' and the maximum (`numeric` or `character`) and minimum (`numeric` or `character`) age which constrain the
```

LewisAJones replied:
Added!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param occdf \code{dataframe}. A dataframe of fossil occurrences. The
+#' dataframe should contain at least three columns: names of taxa (`character`)
+#' and the maximum (`numeric`) and minimum (`numeric`) age which constrain the
+#' age range (in millions of years) of the fossil occurrence. If the age range
```
```suggestion
#' age range of the fossil occurrence. If the age range
```

LewisAJones replied:
I think we should retain this. If the data is not provided in millions of years the plotting functionality will behave oddly!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param occdf \code{dataframe}. A dataframe of fossil occurrences. The
+#' dataframe should contain at least three columns: names of taxa (`character`)
+#' and the maximum (`numeric`) and minimum (`numeric`) age which constrain the
+#' age range (in millions of years) of the fossil occurrence. If the age range
```
Do the values have to be in millions of years? Surely the unit you put in is the unit you get out. I'm just thinking that maybe some people who work with near-recent data might want to use the function. In this case, it might be worth renaming some of the variables, e.g. 'max_ma' -> 'max_time'.

LewisAJones replied:
As above! We can modify this, but would require some fiddling with the plotting functionality (specifically, axis_geo).

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' age range (in millions of years) of the fossil occurrence. If the age range
+#' columns are provided as (`character`) values (i.e., stage names), the
+#' function will attempt to extract numeric age data from the
+#' Geological Timescale by linking stage names.
```
```suggestion
#' Geological Timescale by matching the provided character strings to the stage names.
```

LewisAJones replied:
Updated!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' function will attempt to extract numeric age data from the
+#' Geological Timescale by linking stage names.
+#' @param name \code{character}. The name of the column you wish to be treated
+#' as the input names (e.g., "species" or "genus").
```
```suggestion
#' as the taxonomic names (e.g., "species" or "genus").
```

LewisAJones replied:
Might be worth leaving this more generic. In reality, the user could supply any names here (e.g. species_1)? What do you think?

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param name \code{character}. The name of the column you wish to be treated
+#' as the input names (e.g., "species" or "genus").
+#' @param min_ma \code{character}. The name of the column you wish to be treated
+#' as the input minimum age range (e.g., "min_ma").
```
```suggestion
#' as the minimum limit of the age range (e.g., "min_ma").
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param min_ma \code{character}. The name of the column you wish to be treated
+#' as the input minimum age range (e.g., "min_ma").
+#' @param max_ma \code{character}. The name of the column you wish to be treated
+#' as the input maximum age range (e.g., "max_ma").
```
```suggestion
#' as the maximum limit of the age range (e.g., "max_ma").
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param max_ma \code{character}. The name of the column you wish to be treated
+#' as the input maximum age range (e.g., "max_ma").
+#' @param scale \code{character}. Specify the desired geological timescale to
+#' be used "GTS2020" or "GTS2012". This argument is only useful if the supplied
```
```suggestion
#' be used, either "GTS2020" or "GTS2012". This argument is only useful if the supplied
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param scale \code{character}. Specify the desired geological timescale to
+#' be used "GTS2020" or "GTS2012". This argument is only useful if the supplied
+#' "max_ma" and "min_ma" columns are of class \code{character}.
+#' "GTS2020" is the default.
```
Could link to look_up() here?

LewisAJones replied:
Added, this will likely show up as an error until both branches are merged though!

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+             y = temp_df$taxa_id,
+             pch = 20,
+             col = temp_df$taxa_id)
+      axis_geo(side = 1, intervals = "periods")
```
axis_geo isn't working for me here, although I guess it shouldn't be a problem after merging.

bethany-j-allen replied:
I got there eventually, should be fine as long as deeptime and all other functions in palaeoverse are switched on when the package is loaded.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' are `character` values.
+#'
+#' @return A \code{dataframe} containing the following columns:
+#' unique taxa (`taxa`), taxa ID (`taxa_id`), first appearance of taxa
```
See above re: taxon_id

bethany-j-allen replied:
```suggestion
#' unique taxa (`taxa`), taxa ID (`taxa_id`), first appearance of taxon
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#'
+#' @return A \code{dataframe} containing the following columns:
+#' unique taxa (`taxa`), taxa ID (`taxa_id`), first appearance of taxa
+#' (`FAD_ma`), last appearance of taxa (`LAD_ma`), and temporal range duration
```
```suggestion
#' (`FAD_ma`), last appearance of taxon (`LAD_ma`), and duration of temporal range
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @return A \code{dataframe} containing the following columns:
+#' unique taxa (`taxa`), taxa ID (`taxa_id`), first appearance of taxa
+#' (`FAD_ma`), last appearance of taxa (`LAD_ma`), and temporal range duration
+#' (`range_myr`) is returned.
```
Need to also mention plot and return_error here

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#'
+#' @details The temporal range(s) of taxa are calculated by extracting all
+#' unique taxa (`name` column) from the input `occdf`, and checking their first
+#' and last appearance. The temporal duration of each taxa is also calculated.
```
```suggestion
#' and last appearance. The temporal duration of each taxon is also calculated.
```

bethany-j-allen replied:
I think you need to mention here that these values may be confounding both information about the true duration of taxa and uncertainty in the dating of the occurrences.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  if (class(occdf[, max_ma]) != class(occdf[, min_ma])) {
+    stop("The class of max_ma and min_ma must be the same")
```
```suggestion
    stop("The class of max_ma and min_ma must be the same (`numerical` or `character`")
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  if (any(c(name, min_ma, max_ma) %in% colnames(occdf) == FALSE)) {
+    stop("Either `name`, `min_ma`, or `max_ma`, do not exist in `occdf`")
```
```suggestion
    stop("Either `name`, `min_ma`, or `max_ma`, is not a named column in `occdf`")
```

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+  }
+
+  if (any(is.na(occdf[, name]))) {
+    stop("The `name` column contains NA values")
```
Again, it should maybe be mentioned in the documentation that these need to be filtered out.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+#' @param return_error \code{logical}. Should a vector of numbers be returned
+#' to flag the rows of the `occdf` that cannot be matched to `character`
+#' interval names? This is only relevant if `max_ma` and `min_ma`
+#' are `character` values.
```
Is this needed? I think you could just output NAs for these values in the dataframe and explain that this is the source of the problem.

LewisAJones replied:
I think this provides an easy way for users to identify exactly what rows are the issue. With this vector, they can easily remove or update their data. I think this functionality puts emphasis on the data and that there is an issue they should resolve.

bethany-j-allen commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,178 @@
+        collapse = "\n"
+      ))
+      }
+    }
```
For names, how does this work in practice? NAs aren't allowed, but I could imagine that the PBDB data structure could be a typical one, using 'early_interval' and 'late_interval', but with only 'early_interval' occupied in the case that an occurrence is restricted to a single named interval.

LewisAJones replied:
In this case, the user should supply the min_ma and max_ma as the same (i.e. copy the data across from the early interval to late interval). I agree this requires extra handling, but I think it is good that we don't automate this as it might not be the intention of the user, and it should be flagged.

bethany-j-allen reviewed (commented):
Thanks for putting this together Lewis! I think these will be really fundamental functions in the package. I've put together some thoughts on both the functionality and the documentation. Happy to discuss anything you'd like clarification on or you disagree with.

LewisAJones commented:
Thanks for all the great comments @bethany-j-allen! I've updated tax_range_time for now, and I will try and sort out tax_range_geo next week. Have a great weekend :-)

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+#' This dataframe should contain at least three columns: names of taxa
+#' (`character`), longitude (`numeric`) and latitude (`numeric`).
+#' @param name \code{character}. The name of the column you wish to be treated
+#' as the input names (e.g., "species" or "genus").
```
As before, maybe better to leave this more generic in case the user is not using taxonomic names?

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+#' points in km is returned.
+#' - For the "occ" method, a \code{dataframe} with unique taxa (`taxa`), taxa
+#' ID (`taxa_id`), the number of occupied cells (`cells`), and the spacing
+#' between cells (`spacing`) in km is returned.
```
Integrated.

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+    # Which resolution should be used based on input distance/spacing?
+    # Use the h3jsr::h3_info_table to calculate resolution (however, this
+    # table is not exported in their package, added into palaeoverse)
+    grid <- h3_info_table[
```
I think I've resolved this now, but can you check from your side? It seems to be passing testing and R CMD checks so not sure what the issue might be if it still standing!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,251 @@
+#' - Convex hull: the "con" method calculates the geographic range of a taxa
+#' using a convex hull for each taxa in `occdf`, and calculates the area of
+#' the convex hull (in km\ifelse{html}{\out{<sup>2</sup>}}{\eqn{^2}}) using
+#' \code{\link[geosphere:areaPolygon]{geosphere::areaPolygon()}}.
```
Added a little more details here, but I don't want to go overboard with this! Let me know what you think...

ChristopherDavidDean commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+        h3jsr::point_to_h3(tmp[,c(lng, lat)], res = grid$h3_resolution)
+      )
+      # Calculate number of unique cells occupied
+      oc_df$cells[i] <- length(unique(cells))
```
So the only thing I'm unsure about here is the use of what is being returned. Occupancy isn't super helpful unless there's an understanding of proportion of area occupied - by providing just which cells are occupied you obviously miss this info. Would it be possible to add an additional column to supply this info? You could use the total number of cells which all occurrences appear within as the total number, and use that to propvide a percentage of total for each unique taxa. I think this would be a more useful figure to return to the user. 

Ideally in the future, you'd want to be able to provide extent co-ordinates or some kind of mask to estimate this, but lets save that for a future edition!

LewisAJones replied:
Proportional values added!

ChristopherDavidDean commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+    grid <- h3jsr::h3_info_table[
+      which.min(abs(h3jsr::h3_info_table$avg_cendist_km - spacing)), ]
+    # Add resolution spacing
+    oc_df$spacing <- grid$avg_cendist_km
```
Would it be better to return 'spacing' as a separate object instead of adding it to this dataframe? It seems a bit redundant to just have the same information listed over and over again (I guess it could be useful for further analysis as a category, but it's a bit of a stretch). 

One option would be to return a list, with the dataframe as one entry and the spacing as another. Unsure if this is something that you'd want to implement as it does make getting the dataframe marginally harder, but it would make it a lot more elegant. Just a thought!

LewisAJones replied:
As the number/proportion of occupied cells is hugely dependent on this value, I think it is good for the user to have a column with this data. An alternative would be to have a message returned but this would just get lost if the user is not actively watching the console (i.e. pipeline analysis). I see this as comparable as having a column in your dataframe for quorum level in SQS... ultimately, much easier for a user to drop a column, than to add it manually or handle a list?

ChristopherDavidDean replied:
Yeah, I think now you've added prop. of occupied cells this should be kept. You could have a case where you want to vary the size of grids, and test proportions - in which case you'd want that kept in long format. Happy with this!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+#' @section Developer(s):
+#' Lewis A. Jones
+#' @section Reviewer(s):
+#' Bethany Allen
```
Done!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+#' @importFrom h3jsr point_to_h3
+#' @examples
+#' # Grab internal data
+#' occdf <- tetrapods
```
I don't think so, but I will update it anyway to make it more explicit for users!

LewisAJones replied:
Take that back, this will cause the example to fail unless I assign to another object first. To prevent additional lines being added I've left as is (it works as data is loaded with the package)

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+  }
+
+  if (any(is.na(occdf[,lat])) || any(is.na(occdf[,lng]))) {
+    stop("`lng` and/or `lat` columns contain NA values")
```
Done!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+  } else {
+    method <- possible_methods[method_match]
+  }
+
```
I don't think this is necessary? Unless the method is "occ", spacing isn't used. As implemented, we allow a default for spacing too (100).

LewisAJones commented on man/tax_range_geo.Rd:
```diff
@@ -0,0 +1,130 @@
+% Generated by roxygen2: do not edit by hand
```
Absolutely!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+#' @return A \code{dataframe} with method-specific columns:
+#' - For the "con" method, a \code{dataframe} with unique taxa (`taxa`),
+#' taxon ID (`taxon_id`), convex hull coordinates (`lng` & `lat`), and area
+#' (`area`) in km\ifelse{html}{\out{<sup>2</sup>}}{\eqn{^2}} is returned.
```
I've updated using your first suggestion. I think we've agreed on multiple fronts now that lists are not a desirable output for a lot of people. The return format is also currently in long format which makes it easy to work with tidyverse/ggplot2, etc which I think is desirable for some users.

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+#' - For the "gcd" method, a \code{dataframe} with unique taxa (`taxa`), taxon
+#' ID (`taxon_id`), coordinates of the two most distant points
+#' (`lng` & `lat`), and the 'Great Circle Distance' (`GCD`) between these
+#' points in km is returned.
```
Updated accordingly!

LewisAJones commented on R/tax_range_geo.R:
```diff
@@ -0,0 +1,268 @@
+#' occupied equal-area hexagonal grid cells.
+#'
+#' @param occdf \code{dataframe}. A dataframe of fossil occurrences.
+#' This dataframe should contain at least three columns: names of taxa,
```
I'm not sure this is necessary... but I've added "(see `name`, `lng`, and `lat` arguments)" to appease the reviewer :-D

LewisAJones commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,158 @@
+#' data.
+#'
+#' @param occdf \code{dataframe}. A dataframe of fossil occurrences containing
+#' at least three columns: names of taxa, maximum age and minimum age.
```
Implemented as above. I don't think "names of taxa at a chosen taxonomic level" is required here... in theory you could use any names or type of ID, or have different taxonomic levels.

LewisAJones commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,158 @@
+#' @param plot \code{logical}. Should a plot of the ranges be generated?
+#'
+#' @return A \code{dataframe} containing the following columns:
+#' unique taxa (`taxa`), taxa ID (`taxon_id`), first appearance of taxon
```
This is useful for plotting in this function rather than plotting by taxonomic names. I'd also argue it is easier to see differences between numbers than taxonomic names when they are similar!

LewisAJones commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,158 @@
+#' @importFrom graphics points
+#' @examples
+#' # Grab internal data
+#' occdf <- tetrapods
```
As above!

LewisAJones commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,158 @@
+      segments(x0 = temp_df$max_ma,
+               x1 = temp_df$min_ma,
+               y0 = temp_df$taxon_id,
+               col = temp_df$taxon_id)
```
Updated to only use black to appease the reviewer! :-D

LewisAJones commented on R/tax_range_time.R:
```diff
@@ -0,0 +1,158 @@
+#'
+#' @section Developer(s):
+#' Lewis A. Jones
+#' @section Reviewer(s):
```
Done!

ChristopherDavidDean commented:
Hi Lewis, such a speedy turn around! Merge away, I think it's time to get these functions on the road. Well done everyone!!

willgearty commented:
It looks like there might be an error in one of the tests?

LewisAJones commented:
Hi @ChristopherDavidDean, thank you for all the great comments!

I've implemented the vast majority of your comments, and provided a response if not. 

> It looks like there might be an error in one of the tests, but only with the old version of R?

This is related to the function tax_time_expand, not the functions here. However, let's try and fix this here and then merge. I've removed the specific test that seems to be causing the issue for now.

bethany-j-allen commented:
I can still see a couple of stray 'a taxa' or 'each taxa' that should be 'taxon' in the documentation... But otherwise think this is good to go!
